api/query/atoms_test.go

```diff
@@ -871,10 +890,123 @@ func TestStructuredQuery_bindLink(t *testing.T) {
 		},
 	}
 
+	// AbstractLink should include test-level issues since it's about
+	// searching for links across all runs.
+	// TODO: Shouldn't this also pick up the chromium link too (a.html),
+	// because link isn't meant to care about which product its for?
 	expect := Link{
 		Pattern: "bar",
 		Metadata: map[string][]string{
 			"/testB/b.html": {"bar.com"},
+			"/testC/c.html": {"baz.com"},
+		},
+	}
+	assert.Equal(t, expect, q.BindToRuns(runs...))
+}
+
+func TestStructuredQuery_bindTriaged(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	sha := "sha"
+	mockFetcher := sharedtest.NewMockMetadataFetcher(mockCtrl)
+	mockFetcher.EXPECT().Fetch().Return(&sha, getMetadataTestData(), nil).AnyTimes()
+
+	safari := shared.ParseProductSpecUnsafe("safari")
+	firefox := shared.ParseProductSpecUnsafe("firefox")
+	q := AbstractTriaged{
+		Product:         &firefox,
+		metadataFetcher: mockFetcher,
+	}
+
+	runs := shared.TestRuns{
+		{
+			ID:                int64(0),
+			ProductAtRevision: safari.ProductAtRevision,
+		},
+		{
+			ID:                int64(1),
+			ProductAtRevision: firefox.ProductAtRevision,
+		},
+	}
+
+	expect := Or{
+		Args:[]ConcreteQuery{
+			Triaged{
+				Run: 1,
+				Metadata: map[string][]string{
+					"/testB/b.html": {"bar.com"},
+				},
+			},
+		},
+	}
+	assert.Equal(t, expect, q.BindToRuns(runs...))
+
+	// This case doesn't match, so we should get back False.
+	q = AbstractTriaged{
+		Product:         &safari,
+		metadataFetcher: mockFetcher,
+	}
+	// TODO: This should be False{}, but we pass the entire 'runs' array to
```

Comment:
Right, but like you said these `runs` in the runtime is an array of length 1. So

Reply:
Saying that this 'should be fine' is relying on the implementation detail of `AbstractExists` performing a disjunction over runs. If we want to bake that into `AbstractTriage` (i.e. assume that it is **only** ever called with 1 run), we should add code to error out of `AbstractTriage::BindToRuns` if runs has length >1.

I've done a quick audit, and I think all root atoms only pass single runs into their query atoms. So we could make that change to assume runs is length 1 - WDYT?

Reply:
Yea that is a good idea

Reply:
+1 for the fix and I agree. The search logic in searchcache is quite difficult, and it would require some rethinking to refactor